diffusion/models/stable_diffusion.py

```diff
@@ -142,52 +143,52 @@ def set_rng_generator(self, rng_generator: torch.Generator):
+        if 'attention_mask' in batch:
+            attention_mask = batch['attention_mask']  # mask for text encoders
+            # text mask for U-Net
+            if self.mask_pad_tokens:
+                if len(attention_mask.shape) == 2:
+                    encoder_attention_mask = attention_mask
+                elif len(attention_mask.shape) == 3:
+                    encoder_attention_mask = attention_mask[:, 0]
+                    for i in range(1, attention_mask.shape[1]):
+                        encoder_attention_mask |= attention_mask[:, i]
+                else:
+                    raise ValueError(f'attention_mask should have either 2 or 3 dimensions: {attention_mask.shape}')
+
         # Use latents if specified and available. When specified, they might not exist during eval
         if self.precomputed_latents and self.image_latents_key in batch and self.text_latents_key in batch:
             if self.sdxl:
                 raise NotImplementedError('SDXL not yet supported with precomputed latents')
-            latents, conditioning = batch[self.image_latents_key], batch[self.text_latents_key]
+            latents, text_embeds = batch[self.image_latents_key], batch[self.text_latents_key]
         else:
-            inputs, conditioning = batch[self.image_key], batch[self.text_key]
-            if self.sdxl:
-                # If SDXL, separate the conditioning ([B, 2, 77]) from each tokenizer
-                conditioning, conditioning_2 = conditioning[:, 0, :], conditioning[:, 1, :]
-                conditioning_2 = conditioning_2.view(-1, conditioning_2.shape[-1])
-
-            conditioning = conditioning.view(-1, conditioning.shape[-1])
-            if self.encode_latents_in_fp16:
-                # Disable autocast context as models are in fp16
-                with torch.cuda.amp.autocast(enabled=False):
-                    # Encode the images to the latent space.
-                    # Encode prompt into conditioning vector
-                    latents = self.vae.encode(inputs.half())['latent_dist'].sample().data
-                    if self.sdxl:
-                        conditioning, pooled_conditioning = self.text_encoder([conditioning, conditioning_2])
-                    else:
-                        conditioning = self.text_encoder(conditioning)[0]  # Should be (batch_size, 77, 768)
+            inputs, conditionings = batch[self.image_key], batch[self.text_key]
 
-            else:
-                latents = self.vae.encode(inputs)['latent_dist'].sample().data
-                if self.sdxl:
-                    conditioning, pooled_conditioning = self.text_encoder([conditioning, conditioning_2])
+            # If encode_latents_in_fp16, disable autocast context as models are in fp16
+            c = torch.cuda.amp.autocast(enabled=False) if self.encode_latents_in_fp16 else nullcontext()  # type: ignore
```

Comment:
i'm a bit confused should it be `enabled=not self.encode_latents_in_fp16` or something?

Reply:
TBH, I'm not sure, Mihir added this code, so I was trying to keep the logic the same while condensing it a bit